### PR TITLE
Add mod authentication

### PR DESCRIPTION
**Pull request by Learath2**
I tried to minimise changes

### CODE REVIEW

File: teehistorian/index.php [hunk 1]
```diff
@@ -1,4 +1,50 @@
...
+
+	$hash = null;
+	$salt = null;
+	while(!feof($auths)) {
+		$line = fgets($auths);
+		if($line[0] == '#') {
+			continue;
+		}
+
+		$tokens = explode(" ", $line);
```
Comment (def-) [inline teehistorian/index.php:16]: Are there account names with spaces?

File: teehistorian/index.php [hunk 2]
```diff
@@ -1,4 +1,50 @@
 <html>
+<?php if(isset($_SERVER['PHP_AUTH_USER'])) {
+	$auths = fopen("ddrace_auths.cfg", "r");
+	if(!$auths) {
+		exit;
```
Comment (def-) [inline teehistorian/index.php:5]: could print an error message

File: teehistorian/index.php [hunk 3]
```diff
@@ -1,4 +1,50 @@
 <html>
+<?php if(isset($_SERVER['PHP_AUTH_USER'])) {
+	$auths = fopen("ddrace_auths.cfg", "r");
```
Comment (def-) [inline teehistorian/index.php:3]: Actually, should I put this in the public directory? Not sure if we want this public.

File: teehistorian/index.php [hunk 4]
```diff
@@ -1,4 +1,50 @@
...
+			break;
+		}
+	}
+
+	if(!$hash)
+	{
+		header('WWW-Authenticate: Basic realm="teehistorian"');
+		header('HTTP/1.0 401 Unauthorized');
+		echo "401 Unauthorized";
+		exit;
```
Comment (def-) [inline teehistorian/index.php:33]: This looked weird to me so I checked and apparently this is considered best practice in PHP, oh well: https://www.php.net/manual/en/features.http-auth.php

File: teehistorian/index.php [hunk 5]
```diff
@@ -1,4 +1,50 @@
...
+	$ctx = hash_init("md5");
+	hash_update($ctx, $_SERVER['PHP_AUTH_PW']);
+	hash_update($ctx, hex2bin($salt));
+
+	if($hash !== hash_final($ctx))
+	{
+		header('WWW-Authenticate: Basic realm="teehistorian"');
+		header('HTTP/1.0 401 Unauthorized');
+		echo "401 Unauthorized";
+		exit;
```
Comment (def-) [inline teehistorian/index.php:45]: duplicate from the above, could put into a function.

File: teehistorian/index.php [hunk 6]
```diff
@@ -58,4 +104,10 @@ function formatMiB($size) {
 ?>
 </table>
 <p>Execution time: <?php echo number_format((microtime(true) - $_SERVER["REQUEST_TIME_FLOAT"]) * 1000, 1); ?> ms</p>
+<?php } else {
+	header('WWW-Authenticate: Basic realm="teehistorian"');
+	header('HTTP/1.0 401 Unauthorized');
+	echo "401 Unauthorized";
+	exit;
```
Comment (def-) [inline teehistorian/index.php:111]: triplicate

**Comment by Learath2** [comment]
I was going to fix these, but I think I agree with heinrich that we should
only give out links to demos

**Pull request closed without merging**